h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -177,9 +177,8 @@ private void gridSearch(Grid<MP> grid) {
     try {
       // Get iterator to traverse hyper space
       HyperSpaceWalker.HyperSpaceIterator<MP> it = _hyperSpaceWalker.iterator();
-      // Number of traversed model parameters
-      int counter = grid.getModelCount();
-      while (it.hasNext(model)) {
+      int numberOfBuiltModels = grid.getModelCount();
+      while (it.hasNext(model) && (it.max_models() == 0 || numberOfBuiltModels < it.max_models())) {
```

Comment:
is this a breaking change? for cartesian walker this check did not used to be here

Reply:
@honzasterba for Cartesian max_models is set to a size of the hyperSpace https://github.com/h2oai/h2o-3/blob/39ace2cf92ea362f52cf13ad3d82cfab762e9c6c/h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java#L400

Reply:
May I suggest introducing new method, called e.g. `shouldEarlyStop()`, which would encapsulate early stopping for all the implementations, forcing developers to move the logic into one place ?

Reply:
@Pscheidl i think that the idea of this PR is to move this responsibility outside of our walkers implementations. For `GridSearch` class we don't have hierarchy yet and there is no different implementations in that sense. Are you suggesting to introduce GridSearch interface? 
 I was considering to introduce this method but just to highlight where we do early stopping.

Reply:
No need to do this, I was thinking out loud. Let's skip it in this PR then, as it would only make things more difficult.